torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -1470,6 +1441,82 @@ const Expr* TermExpander::mutate(const RoundOff* v) {
   return term->accept_mutator(this);
 }
 
+Stmt* TermExpander::mutate(const Block* v) {
+  auto vars = v->varBindings();
+  std::vector<Stmt*> stmts;
+  for (Stmt* stmt : *v) {
+    Stmt* stmt_new = stmt->accept_mutator(this);
+    if (stmt_new == nullptr) {
+      continue;
+    }
+
+    if (auto* subBlock = dynamic_cast<Block*>(stmt_new)) {
+      for (auto& pair : subBlock->varBindings()) {
+        vars.emplace_back(pair.first, pair.second);
+      }
+
+      for (Block::iterator I = subBlock->begin(), E = subBlock->end();
+           I != E;) {
+        // Be careful to avoid invalidating the iterator.
+        Stmt* s = *(I++);
+        subBlock->remove_stmt(s);
+        stmts.push_back(s);
+      }
+    } else {
+      stmts.push_back(Stmt::clone(stmt_new));
+    }
+  }
+
+  return new Block(vars, stmts);
+}
+
+Stmt* TermExpander::mutate(const Allocate* v) {
+  const Var* buffer_var_old = v->buffer_var();
+  const Var* buffer_var_new =
+      dynamic_cast<const Var*>(buffer_var_old->accept_mutator(this));
+  bool any_change = buffer_var_new == buffer_var_old;
+
+  const Expr* flattened = getImmediateByType(kInt, 1);
+  std::vector<const Expr*> dims_old = v->dims();
+  std::vector<const Expr*> dims_new(dims_old.size());
+  for (size_t i = 0; i < dims_old.size(); i++) {
+    dims_new[i] = dims_old[i]->accept_mutator(this);
+    any_change |= (dims_new[i] == dims_old[i]);
+    flattened = new Mul(flattened, dims_new[i]);
+  }
+
+  // Safe to do this as there can't be an Allocate inside an Allocate:
+  flattened = IRSimplifier::simplify(flattened);
+
+  if (flattened->isConstant() && immediateEquals(flattened, 0)) {
+    eliminated_allocations_.insert(buffer_var_new);
+    return nullptr;
+  }
+
+  if (!any_change) {
+    return (Stmt*)v;
+  }
+
+  return new Allocate(buffer_var_new, v->dtype(), dims_new);
+}
+
+Stmt* TermExpander::mutate(const Free* v) {
+  const Expr* buffer_var_old = v->buffer_var();
+  const Var* buffer_var_new =
+      dynamic_cast<const Var*>(buffer_var_old->accept_mutator(this));
+
+  if (eliminated_allocations_.count(buffer_var_new)) {
+    eliminated_allocations_.erase(buffer_var_new);
```

Comment:
It would be good if we could make sure that we always delete the corresponding `Free` when we delete an `Allocate`. I understand that it's difficult to do so in IRSimplifier, but maybe we should not be doing it here but rather have it as a part of another pass where we can afford doing a global analysis?

Reply:
Maybe? You'd need to do something like traverse the tree to find allocates you'd want to remove and then for each one traverse the tree for both the allocate and the free. Not sure if that's better but I'd be happy to review a PR if you have a good idea of how that works. At the end of simplification we do check if there is a mismatch between eliminated allocations and frees so we have some amount of safety.